File: src/dataframe/rust/stats/distributions/binomial.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust binomial functions...\n");

  const testCases = [
    { func: "dbinom", distribution: "binomial", args: [3, 10, 0.5, 0] },
    { func: "pbinom", distribution: "binomial", args: [3, 10, 0.5, 1, 0] },
    { func: "qbinom", distribution: "binomial", args: [0.5, 10, 0.5, 1, 0] },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rbinom separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rbinomResult = await callR("rbinom", "binomial", 1, 10, 0.5);
    console.log(`rbinom(10, 0.5) = ${rbinomResult}`);
  } catch (error) {
    console.log(`rbinom: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}

File: packages/dataframe/rust/stats/distributions/binomial.test.R
#!/usr/bin/env Rscript

# Binomial distribution test script
# Usage: Rscript binomial.test.R <function> <x_or_p> <size> <prob> [lower_tail] [log_p]

source("test-helpers.R")

# Define binomial distribution functions
binomial_density <- function(x, size, prob, log = FALSE) {
  dbinom(x, size, prob, log = log)
}

binomial_cumulative <- function(x, size, prob, lower.tail = TRUE, log.p = FALSE) {
  pbinom(x, size, prob, lower.tail = lower.tail, log.p = log.p)
}

binomial_quantile <- function(p, size, prob, lower.tail = TRUE, log.p = FALSE) {
  qbinom(p, size, prob, lower.tail = lower.tail, log.p = log.p)
}

binomial_random <- function(n, size, prob) {
  rbinom(n, size, prob)
}

# Run the test
run_distribution_test(binomial_density, binomial_cumulative, binomial_quantile, binomial_random)
